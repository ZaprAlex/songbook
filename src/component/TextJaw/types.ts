export interface TextJawProps {
    text: string;
    visibleText?: string;
    className?: string;
    wrapperClassName?: string;
    dataTestId?: string;
    onCopyText?: (text: string) => void;
    successMessage?: string;
    inline?: boolean;
}
