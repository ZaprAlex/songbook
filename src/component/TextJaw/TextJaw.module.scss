.text {
    flex-grow: 1;
    font-family: SBSansDisplay-Regular, sans-serif;
    font-size: 18px;
    font-weight: bold;
    line-height: 1.78;
    text-align: center;
    overflow: hidden;
    text-overflow: clip;
    background-image: linear-gradient(to right, rgb(0, 0, 0) 80%, transparent 100%);
    -webkit-background-clip: text;
    color: transparent;

    @media (max-width: 575px) {
        font-size: 16px;
        line-height: normal;
    }

    &-wrapper {
        position: relative;
        display: flex;
        flex-flow: row nowrap;
        justify-content: space-between;
        align-items: center;
        box-sizing: border-box;
        width: 100%;
        padding: 12px;
        background: rgba(0, 0, 0, 0.08);
        border-radius: 8px;
        cursor: pointer;

        @media (max-width: 575px) {
            border-radius: 4px;
        }

        & > img {
            width: 22px;
            height: 22px;
        }
    }
}

.success-msg {
    box-sizing: border-box;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 180px;
    padding: 10px;
    background: #182329;
    border: 1px solid rgba(0, 0, 0, 0.16);
    border-radius: 20px;

    &-wrapper {
        position: absolute;
        top: -105%;
        left: 0;
        width: 100%;
        display: flex;
        justify-content: center;
        align-items: center;
    }

    & > span {
        margin: 0 10px;
        font-family: SBSansDisplay-Regular, sans-serif;
        font-size: 12px;
        font-weight: bold;
        line-height: 1.56;
        text-align: center;
        vertical-align: center;
        color: #ffffff;
    }
}

.inline {
    white-space: nowrap;
    padding: 0 8px 0 26px;
}
