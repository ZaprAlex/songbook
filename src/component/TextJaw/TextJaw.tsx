import React, { useEffect, useRef, useState } from 'react';
import cn from 'classnames';

import { sleep } from '../../utils/helper';
import { copyToClipboard } from '../../utils/copyToClipboard';
import copyIcon from '../../static/svgs/copy-icon.svg';

import { TextJawProps } from './types';
import styles from './TextJaw.module.scss';

const TextJaw: React.FC<TextJawProps> = ({
    text,
    visibleText = text,
    dataTestId = 'text-jaw',
    className = '',
    wrapperClassName,
    onCopyText,
    successMessage,
    inline = true,
}) => {
    const textRef = useRef<HTMLDivElement>(null);
    const [show, setShow] = useState(false);

    useEffect(() => {
        let isMounted = true;

        (async function () {
            await sleep(2);
            isMounted && setShow(false);
        })();

        return () => {
            isMounted = false;
        };
    }, [show]);

    const copyText = () => {
        if (text) {
            copyToClipboard(text, () => {
                setShow(true);
                onCopyText?.(text);
            });
        }
    };

    return (
        <div className={cn(styles.textWrapper, wrapperClassName)} onClick={copyText} data-test-id={dataTestId}>
            <p className={cn(styles.text, className, { [styles.inline]: inline })} ref={textRef}>
                {visibleText}
            </p>
            <img src={copyIcon} alt="" />
            {show && successMessage && (
                <div className={styles.successMsgWrapper} data-test-id="success-modal">
                    <div className={styles.successMsg}>
                        <span>{successMessage}</span>
                    </div>
                </div>
            )}
        </div>
    );
};

export default TextJaw;
