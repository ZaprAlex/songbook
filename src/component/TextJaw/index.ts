export { default } from './TextJaw';
export * from './types';
