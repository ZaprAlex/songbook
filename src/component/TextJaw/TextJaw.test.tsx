import React from 'react';
import { render, screen } from '@testing-library/react';

import TextJaw from './TextJaw';

describe('TextJaw', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.restoreAllMocks();
    });

    test('snap default', () => {
        const props = { text: 'PROMO_CODE' };

        const { container } = render(<TextJaw {...props} />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('snap with props', () => {
        const onClick = jest.fn();
        const dataTestId = 'promo-code-jaw';
        const successMessage = 'Промокод скопирован';
        const props = { text: 'PROMO_CODE', successMessage, onClick, dataTestId };

        render(<TextJaw {...props} />);

        expect(screen.getByTestId(dataTestId)).toBeInTheDocument();
    });
});
