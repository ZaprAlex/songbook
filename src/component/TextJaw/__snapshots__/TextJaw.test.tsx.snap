// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextJaw snap default 1`] = `
<div
  class="textWrapper"
  data-test-id="text-jaw"
>
  <p
    class="text inline"
  >
    PROMO_CODE
  </p>
  <img
    alt=""
    src="copy-icon.svg"
  />
</div>
`;
